### PR TITLE
feat:(install) Support JFrog Artifactory

### PR DESCRIPTION
**Pull request by yahavi**
This pull request replaces these pull requests.
https://github.com/jenkins-x/jx/pull/1274
https://github.com/jenkins-x/jx/pull/1384
https://github.com/jenkins-x/draft-packs/pull/50
https://github.com/jenkins-x/jenkins-x-image/pull/32

This PR, unlike the ones it replaces, does not include any changes to the draft-packs. The default draft-packs work properly with Artifactory according to our testing.
This simplifies the implementation and I hope can help accepting this functionality faster.
Please let me know what you think.
Thanks

### CODE REVIEW

File: pkg/jx/cmd/install.go [hunk 1]
```diff
@@ -1179,3 +1194,55 @@ func (options *InstallOptions) dockerRegistryValue() (string, error) {
...
+	if options.AdminSecretsService.Flags.ArtifactoryUrl != "" {
+		return nil, ArtifactoryRepoMgr
+	}
+	binaryRepositoryManagers := []string{NexusRepoMgr, ArtifactoryRepoMgr}
+	prompts := &survey.Select{
+		Message: "Select binary repository manager:",
+		Options: binaryRepositoryManagers,
+		Default: NexusRepoMgr,
+		Help:    "Binary repository manager to resolve dependencies and for uploading builds",
+	}
```
Comment (rawlingsj) [inline pkg/jx/cmd/install.go]: We should probably make it clear that artifactory is assumed to already be running outside the cluster somewhere, ideally instead we'd optionally install and configure it incluster if folks want.  Maybe for now we just make it clear that Nexus is installed while an existing Artifactory is needed?

**Comment by rawlingsj** [comment]
Indeed this does look much simpler - thanks.  Right now our BDD tests that run on this PR do not install Nexus so we need to manually verify this doesn't break any existing default paths.

As an aside we should probably auto install Artifactory which also sets the default admin password the same as we do nexus if someone chooses Artifactory from the wizard.  Perhaps for now we just make the wizard a bit clearer.  I added a review comment.

**Comment by yahavi** [comment]
@rawlingsj, Thanks for the review.
We are planning to add Artifactory in-cluster in the future.
For now, to make the wizard a bit clearer, I changed the "Artifactory" option to "External Artifactory".

File: pkg/jx/cmd/install.go (hunk 1 above)
Comment (yahavi) [reply pkg/jx/cmd/install.go]: Changed "Artifactory" --> "External Artifactory".

**Comment by yahavi** [comment]
@rawlingsj & @ccojocar, any progress with this?

**Comment by yahavi** [comment]
@ccojocar,
According to @rawlingsj [comment](https://github.com/jenkins-x/jx/pull/1753#issuecomment-423799016), I can ignore the bdd test results. Right?

**Comment by eyalbe4** [comment]
@ccojocar and @rawlingsj,

Since validating this PR seems to be taking a very long time (the original discussion around this started back in July), I was wondering if there's anything else we can do to help.
Can you please share what else needs to be done before this PR can be merged?

Appreciate your help with this.

**Comment by jstrachan** [comment]
I wonder if you could rebase to fix the conflicts; hopefully we can get this merged soon

**Comment by yahavi** [comment]
> I wonder if you could rebase to fix the conflicts; hopefully we can get this merged soon

@jstrachan - I fixed the conflicts now.

**Comment by eyalbe4** [comment]
@jstrachan / @ccojocar / @rawlingsj,
Any progress with approving this PR?

**Comment by yahavi** [comment]
@jstrachan / @ccojocar / @rawlingsj,
I haven't heard back from you for a long time. The fact that this PR remains idle forces me to continuously fix conflicts.
Any reason for the set back?
Do you see any issues with merging this?

**Comment by rawlingsj** [comment]
Yeah sorry, I know it's bad form to let a PR go stale like this.  We had a lot of changes in a small amount of time which meant we were having to work hard to keep stable.  My main concern for this ATM is there's no automated testing.  As soon as this is merged we'd have to maintain it and we don't have the infrastructure setup to do so.  All our automated BDD tests work using the Nexus path so really we'd need to do the same for Artifactory.  

We'll need to prioritise work in the new year to setup our testing infra and get some automated BDD tests done.

**Comment by yahavi** [comment]
@rawlingsj Can you share some information about what's missing in terms of tests infrastructure? Either me or someone else in the community may be able to help.

**Comment by ccojocar** [comment]
@jstrachan @rawlingsj @pmuir How are we going to move forward with this? Should the JFrog Artifactory be converted into a jx `app`? It seems like a good fit to me. 

@yahavi @eyalbe4 Would you be interested to create an application? 

@pmuir Please could you point to some documentation? Thanks

**Comment by eyalbe4** [comment]
@ccojocar,
Can you please elaborate more about the suggested solution? What does using "jx app" mean?
We understand that the issue blocking the merge of this PR is the lack of testing infrastructure (which we're willing to help set up).

**Comment by ccojocar** [comment]
@eyalbe4 These are some apps examples https://github.com/jenkins-x-apps which you can use to create an application for Artifactory. Also here you can find some docs https://jenkins-x.io/apps/. Please join our slack channel, if you have any questions, @pmuir will help you to build the app.

@yahavi The install has changed significantly since this PR was opened. Do you agree to close it, and follow up with a different PR based on the current code when the app is ready?

**Comment by StevenACoffman** [comment]
@yahavi @eyalbe4 Did this ever get picked up again?

**Comment by eyalbe4** [comment]
@StevenACoffman,
As far as I know, no one took this initiative following our attempt to push this functionality.

**Comment by StevenACoffman** [comment]
https://jfrog.com/blog/jenkins-x-code-to-production-in-high-speed/

**Comment by cdoughty-r7** [comment]
From above link:
> Note: The Artifactory Jenkins X enhancement has not yet been merged into the official Jenkins X GitHub repository. JFrog is working together with CloudBees to merge the Support JFrog Artifactory pull request.

This just links back to these PRs: https://github.com/jenkins-x/jx/pull/1753 / https://github.com/jenkins-x/jx/pull/1384

**Pull request closed without merging**